File: src/algorithms/pathfinding/dijkstra/sources/Dijkstra.java
import java.util.*;

public class Dijkstra {
    public static int[][] dijkstra(int[][] grid, int[] start, int[] end) {
        int rowCount = grid.length;
        int colCount = grid[0].length;
        int[][] distance = new int[rowCount][colCount];
        for (int[] row : distance) Arrays.fill(row, Integer.MAX_VALUE);
        distance[start[0]][start[1]] = 0;
        int[][][] parent = new int[rowCount][colCount][];
        boolean[][] visitedSet = new boolean[rowCount][colCount];

        PriorityQueue<int[]> openSet = new PriorityQueue<>(
            Comparator.comparingInt(nodeA -> nodeA[2])
        );
        openSet.add(new int[]{start[0], start[1], 0});

        int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

        while (!openSet.isEmpty()) {
            int[] current = openSet.poll();
            int currentRow = current[0];
            int currentCol = current[1];
            if (visitedSet[currentRow][currentCol]) continue;
            visitedSet[currentRow][currentCol] = true;

            if (currentRow == end[0] && currentCol == end[1]) {
                return reconstructPath(parent, end);
            }

            for (int[] direction : directions) {
                int neighborRow = currentRow + direction[0];
                int neighborCol = currentCol + direction[1];
                if (neighborRow < 0 || neighborRow >= rowCount) continue;
                if (neighborCol < 0 || neighborCol >= colCount) continue;
                if (grid[neighborRow][neighborCol] == 1) continue;
                if (visitedSet[neighborRow][neighborCol]) continue;

                int newDistance = distance[currentRow][currentCol] + 1;
                if (newDistance < distance[neighborRow][neighborCol]) {
                    distance[neighborRow][neighborCol] = newDistance;
                    parent[neighborRow][neighborCol] = new int[]{currentRow, currentCol};
                    openSet.add(new int[]{neighborRow, neighborCol, newDistance});
                }
            }
        }

        return new int[0][];
    }

    private static int[][] reconstructPath(int[][][] parent, int[] end) {
        List<int[]> path = new ArrayList<>();
        int[] current = end;
        while (current != null) {
            path.add(0, current);
            current = parent[current[0]][current[1]];
        }
        return path.toArray(new int[0][]);
    }
}

File: src/algorithms/pathfinding/dijkstra/dijkstra.ts
import type { GridCell } from "@/types";

export interface DijkstraResult {
  path: [number, number][];
  visited: [number, number][];
}

export function dijkstra(
  grid: GridCell[][],
  start: [number, number],
  end: [number, number],
): DijkstraResult {
  const rowCount = grid.length;
  const colCount = grid[0]?.length ?? 0;
  const visited: [number, number][] = [];

  /* Distance map: distance[row][col] = shortest known distance from start */
  const distance: number[][] = Array.from({ length: rowCount }, () =>
    new Array(colCount).fill(Infinity),
  );
  distance[start[0]]![start[1]] = 0;

  /* Parent map for path reconstruction */
  const parent: ([number, number] | null)[][] = Array.from({ length: rowCount }, () =>
    new Array<[number, number] | null>(colCount).fill(null),
  );

  /* Simple priority queue backed by a sorted array */
  const openSet: { row: number; col: number; dist: number }[] = [
    { row: start[0], col: start[1], dist: 0 },
  ];

  const visitedSet: boolean[][] = Array.from({ length: rowCount }, () =>
    new Array(colCount).fill(false),
  );

  const directions: [number, number][] = [
    [-1, 0],
    [1, 0],
    [0, -1],
    [0, 1],
  ];

  while (openSet.length > 0) {
    /* Sort by distance ascending and take the closest node */
    openSet.sort((nodeA, nodeB) => nodeA.dist - nodeB.dist);
    const current = openSet.shift()!;

    if (visitedSet[current.row]![current.col]) continue;
    visitedSet[current.row]![current.col] = true;
    visited.push([current.row, current.col]);

    /* Check if we reached the end */
    if (current.row === end[0] && current.col === end[1]) {
      const path = reconstructPath(parent, end);
      return { path, visited };
    }

    /* Explore 4-directional neighbors */
    for (const [deltaRow, deltaCol] of directions) {
      const neighborRow = current.row + deltaRow;
      const neighborCol = current.col + deltaCol;

      if (
        neighborRow < 0 ||
        neighborRow >= rowCount ||
        neighborCol < 0 ||
        neighborCol >= colCount
      ) {
        continue;
      }

      const neighborCell = grid[neighborRow]![neighborCol]!;
      if (neighborCell.type === "wall") continue;
      if (visitedSet[neighborRow]![neighborCol]) continue;

      const newDistance = distance[current.row]![current.col]! + 1;
      if (newDistance < distance[neighborRow]![neighborCol]!) {
        distance[neighborRow]![neighborCol] = newDistance;
        parent[neighborRow]![neighborCol] = [current.row, current.col];
        openSet.push({ row: neighborRow, col: neighborCol, dist: newDistance });
      }
    }
  }

  /* No path found */
  return { path: [], visited };
}

function reconstructPath(
  parent: ([number, number] | null)[][],
  end: [number, number],
): [number, number][] {
  const path: [number, number][] = [];
  let current: [number, number] | null = end;

  while (current !== null) {
    path.unshift(current);
    current = parent[current[0]]![current[1]]!;
  }

  return path;
}
